autoeq/frequency_response.py

```diff
@@ -247,6 +260,70 @@ def write_to_csv(self, file_path=None):
         df = pd.DataFrame(self.to_dict())
         df.to_csv(file_path, header=True, index=False, float_format='%.2f')
 
+    def write_supereq(self, file_path):  # interpolation method = scipy.interpolate.krogh_interpolate()
+        """Writes SuperEq's 18 band eq settings to a preset file."""
+        """Gain values are interpolated using Krogh's interpolator method from same memory data which is stored into csv file by AutoEQ"""
+        last_idx = np.size(self.frequency)-1
+        prev_idx = 0
+        self.gains = []  # array slot used in plot_graph()
+        with open(file_path.replace('.txt', '.feqk'), 'w', encoding='utf-8') as f:  # .feq = foobar2k preset file extension
+            s = ''
+            file_path = os.path.abspath(file_path)
+            cfreq = [55.0, 77.0, 110.0, 156.0, 220.0, 311.0, 440.0, 622.0, 880.0, 1200.0, 1800.0, \
+                     2500.0, 3500.0, 5000.0, 7000.0, 10000.0, 14000.0, 20000.0] # SuperEQ band center frequencies
+            idx = 0
+            for i in cfreq:
+                idx = bisect.bisect_left(self.frequency, i, lo=prev_idx, hi=last_idx-1)
+                if last_idx > idx+1:
+                    P = krogh_interpolate([self.frequency[idx-2], self.frequency[idx-1], self.frequency[idx], self.frequency[idx+1]], \
```

Comment:
I did not want to translate SuperEQ sources to Python language so, as you have the **equalization** curve data already solved there it's easiest to just interpolate each (18) bandwidth for gain at bandwidths center frequency. AFAIK, there is not much transition overflow between bands in SuperEQ filters (test with sweep signal shows little Gibbs phenomena type effect in output) and, if there is then isn't it just adjusting little bit final band gain then (I can try solve how much it effects in gains of the bands). 

Of course one could try use those sub classes too but it may become more complex then... .

No problem to move interpolation to function of its own.

Reply:
I don't think you need to implement the entire SuperEQ. If I'm not mistaken, SuperEQ filters simply look like bandpass filters (though there's probably a different term, since SuperEQ's filter would have 0 dB for everything, and a specific gain for a given band):

![bandpass filter](https://upload.wikimedia.org/wikipedia/commons/6/6b/Bandwidth_2.svg)

Such a filter, I guess, can be thought of as a low shelf at $f_H$ minus a low shelf at $f_L$ (or high shelf at $f_L$ minus a high shelf at $f_H$). Low and high shelfs are already there, you just need to tie their gains.

Reply:
Hmm... but you still don't get the right gain value because of same reason as what it is with peak filters. Here's an example bi-quad LS and HS constructed as you say (with gains +3dB, Q=0.7 and bandwidths by the SuperEQ specs) trying to follow straight line at 3dB level:

![combinedlshl](https://github.com/jaakkopasanen/AutoEq/assets/141587285/c3af4a8b-b5a9-45cc-843c-14b5002c27a8)

_Red dashed curve is combined response, black dashed curve is one peaking filter with same design properties._

As you see ,only first and last filter exceeds target 3dB gain at center frequency ... all others have little bit lower peak gain at center frequency (~2.7dB). Now, if you make combined response follow the red target 3dB curve you need to drop the design gains of most of the filters to much lower values... and these values would be your SuperEQ gain values. Implementation (interpolation of equalization curve), which I suggest, measures directly to the target 3dB curve so I get 3dB as what is expected (if we forget Gibbs phenomena type phenomena band neighboring frequency areas).:

![foobarfilter](https://github.com/jaakkopasanen/AutoEq/assets/141587285/70399328-3574-45e4-afff-99f26b52ee17)

Example: Sweep signal filtered by Foobar2000 EQ vs target response.